Accomodate Reduction IterDomains when concretizing reshape extents

jacobhinkle opened the pull request:
We register extents for concretization when we concretize reshape. In order to do that, we line up `IterDomain`s in the symbolic reshaped TV and the new, concretized one. In cases where the concretized reshape is trivial, such as when the output shape is the same as the input, we do not create a new TV. In those cases, we will have the input to the original `ViewOp` as the concretized output. That input TV might have reduction domains, as in the provided test, in which case we need to filter those out when doing this alignment. This small PR just implements that filtering.

Fixes #1691.

jacobhinkle commented:
!build

naoyam commented:
> We register extents for concretization when we concretize reshape. In order to do that, we line up `IterDomain`s in the symbolic reshaped TV and the new, concretized one. In cases where the concretized reshape is trivial, such as when the output shape is the same as the input, we do not create a new TV. In those cases, we will have the input to the original `ViewOp` as the concretized output. That input TV might have reduction domains, as in the provided test, in which case we need to filter those out when doing this alignment. This small PR just implements that filtering.
> 
> Fixes #1691.

Could you give a concrete example? 

> In cases where the concretized reshape is trivial, such as when the output shape is the same as the input, we do not create a new TV. In those cases, we will have the input to the original `ViewOp` as the concretized output. 

This is a bit unexpected for me. Are you saying the reshape would look like: `tv0 = ViewOp(tv0, ...)`? That seems like violating the SSA assumption in the Fusion IR.

jacobhinkle commented:
> a concrete example

Ah, sorry I should've been more clear. Say we have a dynamic reshape like
```
Inputs:
  T0
Outputs:
  T2
T1[ ?S2{i2} ?S3{i3} ] = view(T0[ iS0{i0} iS1{i1} ])
T2[ ?S4{i2} ?S5{i3} ] = -T1[ ?S2{i2} ?S3{i3} ]
```
Then if it turns out that `i2==i0` and `i3==i1`, then this is actually concretized as
```
Inputs:
  T0
Outputs:
  T2
T2[ iS4{i2} iS5{i3} ] = -T0[ iS0{i0} iS1{i1} ]
```
What happened is we replaced `T1[ ?S2{i2} ?S3{i3} ]` with `T0[ iS0{i0} iS1{i1} ]` in all uses. We did not just replace the definition of `T1` with `set(T0)`.

Now imagine we have instead
```
Inputs:
  T0
Outputs:
  T3
T1[ iS2{i0} rS3{i1} ] = sum(T0[ iS0{i0} iS1{i1} ])
T2[ ?S4{i2} ] = view(T1[ iS2{i0} rS3{i1} ])
T3[ ?S4{i2} ] = -T2[ ?S4{i2} ]
```
This is like the case in the issue/test. Now, when we concretize we have
```
Inputs:
  T0
Outputs:
  T3
T1[ iS2{i0} rS3{i1} ] = sum(T0[ iS0{i0} iS1{i1} ])
T3[ iS4{i2} ] = -T1[ iS2{i0} rS3{i1} ]
```
Here we have replaced `T2[ ?S4{i2} ]` with `T1[ iS2{i0} rS3{i1} ]`.

naoyam commented:
Thanks. Could you add that to the code?

naoyam commented on csrc/dynamic_transform.cpp:
```diff
@@ -647,8 +647,10 @@ void DynamicTransformConcretizer::concretizeReshape() {
     // like it would have if we had used a static reshape instead.
-    auto old_rfactor = incomplete_out_tv->getMaybeRFactorDomain();
-    auto new_rfactor = concrete_reshape_out_tv->getMaybeRFactorDomain();
+    auto old_rfactor =
+        TensorDomain::noReductions(incomplete_out_tv->getMaybeRFactorDomain());
+    auto new_rfactor = TensorDomain::noReductions(
+        concrete_reshape_out_tv->getMaybeRFactorDomain());
```
Now I understand `concrete_reshape_out_tv` can have reduction domains. When is that also the case with `incomplete_out_tv`?

jacobhinkle replied:
That should not happen, since that domain is always created by the `reshape` op. I'll remove that filter.

naoyam reviewed (approved):
LGTM. Thanks for the fix.
